lib/OpenQA/WebAPI/Controller/API/V1/Search.pm

```diff
@@ -0,0 +1,114 @@
+# Copyright (C) 2020 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::WebAPI::Controller::API::V1::Search;
+use Mojo::Base 'Mojolicious::Controller';
+
+use OpenQA::Utils;
+use Mojo::File 'path';
+use IPC::Run;
+
+=pod
+
+=head1 NAME
+
+OpenQA::WebAPI::Controller::API::V1::Search
+
+=head1 SYNOPSIS
+
+  use OpenQA::WebAPI::Controller::API::V1::Search;
+
+=head1 DESCRIPTION
+
+Implements API methods to search (for) tests.
+
+=head1 METHODS
+
+=cut
+
+# Helper methods
+
+=over 4
+
+=item query()
+
+Renders an array of search results.
+
+=over 8
+
+=item q
+
+A YAML document describing the job template. The template will be validated against the schema.
+
+=back
+
+The response will have these fields:
+
+=over
+
+=item B<data>: the array of search results, with B<occurrence> and B<contents> in case of a fulltext match
+
+=item B<error>: an array of errors if validation or retrieving results failed
+
+=back
+
+The B<data> and B<error> fields are mutually exclusive.
+
+=back
+
+=cut
+
+sub query {
```

Comment:
This route seems to use a lot of resources. It spawns a sub process, does a lot of file system IO, keeps potentially a lot of results in memory and so on. Maybe it would make sense to restrict using that route to only a limited number of usages per minute. That's for instance how some forums do it. (I have no idea whether Mojolicious supports something like that. @kraih ) We could also de-prioritize the `git grep` sub process (especially IO-wise) but I'd also had to look how to do it.

I'd just like to note that openQA already feels to slow so we should avoid keeping the web UI host even more busy.

Reply:
Mojolicious has nothing built-in for this. If you just want simple rate limiting you could use a Minion lock though.
```
# Only allow about 5 queries per Minute
return $self->render(text => 'Rate limit!') unless $self->app->minion->lock('webui_query_rate_limit', 60, {limit => 5});
```

Reply:
Using a Minion lock sounds good and actually quite simple to use. Of course I meant to limit the route usage per user (so the lock name should contain the user name).